### DISCUSSION TITLE
inspect.lua is now written in Teal

### DISCUSSION THREAD

**Discussion by kikito** [teal-language/tl#509] [Show and tell]
Hello Teal Community! (tealers?, tealors?)

I just wanted to announce that I rewrote my most popular library, [inspect.lua](https://github.com/kikito/inspect.lua), to Teal.

The process was very straightforward - I believe the most difficult step was using 3-space indentations instead of my usual 2. My library is unusual in that it has to be able to deal with `any` types (it accepts any Lua value). This turned out to not be a huge deal - I just needed to use `as` at some points.

I also included some performance-related improvements.

These changes are all backwards-compatible with the library - if you are already using inspect.lua, it should be safe to upgrade. I have some future non-backwards compatible changes in mind.

**Comment by catwell** [comment]
> I believe the most difficult step was using 3-space indentations instead of my usual 2.

Why did you need to do that? That was unrelated to Teal, right?

**Reply by lenscas** [reply]
just to be sure, I copied the `inspect.tl` file from his repo into the teal playground and started to indent stuff with random amounts.

The playground at least seems to always compile it no matter what indentation is used. However, the indentation used by the generated lua is using 3 spaces regardless of the amount of spaces used in the teal file. The generated lua code also happily indents stuff when there was no indentation before if it thinks there should be.

**Reply by hishamhm** [reply]
Yes, Teal does not enforce indentation for its input at all (the input grammar is very Lua-like in this aspect, quirks and all).

The generated code is indented as @lenscas described (mostly for debugability of the compiler, initially!), so it does not preserve the original horizontal layout of the input, but it it tries hard to keep the _vertical_ layout (line breaks) as close as possible to the original, just so that runtime errors generate stack traces with line numbers matching the Teal input.

**Reply by kikito** [reply]
It was related with Teal in the sense that its Lua output is indented by 3 spaces.

I wanted to diff the pre-Teal `inspect.lua` code with the post-Teal generated `inspect.lua` code, in order to easily spot what changed, see inconsistencies, and the like.

Using my editor to retab was the less painful way for me to do that. I'm almost certain there's a flag somewhere in git to make this, but I am never able to memorize such things and digging into git's inconsistent console UI makes me angry.

**Comment by nagolove** [comment]
Great news, thanks! I glad that you do not abandon this useful project. But I have did my own instect.d.tl wrapper for inspect.lua 😀
